<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang xml:lang>
<head>
  <meta charset="utf-8" />
  <meta name="generator" content="pandoc" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
  <title>portal-api</title>
  <style type="text/css">
      code{white-space: pre-wrap;}
      span.smallcaps{font-variant: small-caps;}
      span.underline{text-decoration: underline;}
      div.column{display: inline-block; vertical-align: top; width: 50%;}
  </style>
  <style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; position: absolute; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; }
pre.numberSource a.sourceLine:empty
  { position: absolute; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: absolute; left: -5em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; } /* Alert */
code span.an { color: #008000; } /* Annotation */
code span.at { } /* Attribute */
code span.bu { } /* BuiltIn */
code span.cf { color: #0000ff; } /* ControlFlow */
code span.ch { color: #008080; } /* Char */
code span.cn { } /* Constant */
code span.co { color: #008000; } /* Comment */
code span.cv { color: #008000; } /* CommentVar */
code span.do { color: #008000; } /* Documentation */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.im { } /* Import */
code span.in { color: #008000; } /* Information */
code span.kw { color: #0000ff; } /* Keyword */
code span.op { } /* Operator */
code span.ot { color: #ff4000; } /* Other */
code span.pp { color: #ff4000; } /* Preprocessor */
code span.sc { color: #008080; } /* SpecialChar */
code span.ss { color: #008080; } /* SpecialString */
code span.st { color: #008080; } /* String */
code span.va { } /* Variable */
code span.vs { color: #008080; } /* VerbatimString */
code span.wa { color: #008000; font-weight: bold; } /* Warning */
  </style>
  <style type="text/css">@font-face {
font-family: octicons-link;
src: url(data:font/woff;charset=utf-8;base64,d09GRgABAAAAAAZwABAAAAAACFQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABEU0lHAAAGaAAAAAgAAAAIAAAAAUdTVUIAAAZcAAAACgAAAAoAAQAAT1MvMgAAAyQAAABJAAAAYFYEU3RjbWFwAAADcAAAAEUAAACAAJThvmN2dCAAAATkAAAABAAAAAQAAAAAZnBnbQAAA7gAAACyAAABCUM+8IhnYXNwAAAGTAAAABAAAAAQABoAI2dseWYAAAFsAAABPAAAAZwcEq9taGVhZAAAAsgAAAA0AAAANgh4a91oaGVhAAADCAAAABoAAAAkCA8DRGhtdHgAAAL8AAAADAAAAAwGAACfbG9jYQAAAsAAAAAIAAAACABiATBtYXhwAAACqAAAABgAAAAgAA8ASm5hbWUAAAToAAABQgAAAlXu73sOcG9zdAAABiwAAAAeAAAAME3QpOBwcmVwAAAEbAAAAHYAAAB/aFGpk3jaTY6xa8JAGMW/O62BDi0tJLYQincXEypYIiGJjSgHniQ6umTsUEyLm5BV6NDBP8Tpts6F0v+k/0an2i+itHDw3v2+9+DBKTzsJNnWJNTgHEy4BgG3EMI9DCEDOGEXzDADU5hBKMIgNPZqoD3SilVaXZCER3/I7AtxEJLtzzuZfI+VVkprxTlXShWKb3TBecG11rwoNlmmn1P2WYcJczl32etSpKnziC7lQyWe1smVPy/Lt7Kc+0vWY/gAgIIEqAN9we0pwKXreiMasxvabDQMM4riO+qxM2ogwDGOZTXxwxDiycQIcoYFBLj5K3EIaSctAq2kTYiw+ymhce7vwM9jSqO8JyVd5RH9gyTt2+J/yUmYlIR0s04n6+7Vm1ozezUeLEaUjhaDSuXHwVRgvLJn1tQ7xiuVv/ocTRF42mNgZGBgYGbwZOBiAAFGJBIMAAizAFoAAABiAGIAznjaY2BkYGAA4in8zwXi+W2+MjCzMIDApSwvXzC97Z4Ig8N/BxYGZgcgl52BCSQKAA3jCV8CAABfAAAAAAQAAEB42mNgZGBg4f3vACQZQABIMjKgAmYAKEgBXgAAeNpjYGY6wTiBgZWBg2kmUxoDA4MPhGZMYzBi1AHygVLYQUCaawqDA4PChxhmh/8ODDEsvAwHgMKMIDnGL0x7gJQCAwMAJd4MFwAAAHjaY2BgYGaA4DAGRgYQkAHyGMF8NgYrIM3JIAGVYYDT+AEjAwuDFpBmA9KMDEwMCh9i/v8H8sH0/4dQc1iAmAkALaUKLgAAAHjaTY9LDsIgEIbtgqHUPpDi3gPoBVyRTmTddOmqTXThEXqrob2gQ1FjwpDvfwCBdmdXC5AVKFu3e5MfNFJ29KTQT48Ob9/lqYwOGZxeUelN2U2R6+cArgtCJpauW7UQBqnFkUsjAY/kOU1cP+DAgvxwn1chZDwUbd6CFimGXwzwF6tPbFIcjEl+vvmM/byA48e6tWrKArm4ZJlCbdsrxksL1AwWn/yBSJKpYbq8AXaaTb8AAHja28jAwOC00ZrBeQNDQOWO//sdBBgYGRiYWYAEELEwMTE4uzo5Zzo5b2BxdnFOcALxNjA6b2ByTswC8jYwg0VlNuoCTWAMqNzMzsoK1rEhNqByEyerg5PMJlYuVueETKcd/89uBpnpvIEVomeHLoMsAAe1Id4AAAAAAAB42oWQT07CQBTGv0JBhagk7HQzKxca2sJCE1hDt4QF+9JOS0nbaaYDCQfwCJ7Au3AHj+LO13FMmm6cl7785vven0kBjHCBhfpYuNa5Ph1c0e2Xu3jEvWG7UdPDLZ4N92nOm+EBXuAbHmIMSRMs+4aUEd4Nd3CHD8NdvOLTsA2GL8M9PODbcL+hD7C1xoaHeLJSEao0FEW14ckxC+TU8TxvsY6X0eLPmRhry2WVioLpkrbp84LLQPGI7c6sOiUzpWIWS5GzlSgUzzLBSikOPFTOXqly7rqx0Z1Q5BAIoZBSFihQYQOOBEdkCOgXTOHA07HAGjGWiIjaPZNW13/+lm6S9FT7rLHFJ6fQbkATOG1j2OFMucKJJsxIVfQORl+9Jyda6Sl1dUYhSCm1dyClfoeDve4qMYdLEbfqHf3O/AdDumsjAAB42mNgYoAAZQYjBmyAGYQZmdhL8zLdDEydARfoAqIAAAABAAMABwAKABMAB///AA8AAQAAAAAAAAAAAAAAAAABAAAAAA==) format('woff');
}
body {
-webkit-text-size-adjust: 100%;
text-size-adjust: 100%;
color: #333;
font-family: "Helvetica Neue", Helvetica, "Segoe UI", Arial, freesans, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
font-size: 16px;
line-height: 1.6;
word-wrap: break-word;
width: 925px;
max-width: 99%;
box-sizing: border-box;
padding: 30px 30px 8rem 30px;
margin-left: auto;
margin-right: auto;
}
body a {
background-color: transparent;
}
body a:active,
body a:hover {
outline: 0;
}
body strong {
font-weight: bold;
}
body h1 {
font-size: 2em;
margin: 0.67em 0;
}
body img {
border: 0;
}
body hr {
box-sizing: content-box;
height: 0;
}
body pre {
overflow: auto;
}
body code,
body kbd,
body pre {
font-family: monospace, monospace;
font-size: 1em;
}
body input {
color: inherit;
font: inherit;
margin: 0;
}
body html input[disabled] {
cursor: default;
}
body input {
line-height: normal;
}
body input[type="checkbox"] {
box-sizing: border-box;
padding: 0;
}
body table {
border-collapse: collapse;
border-spacing: 0;
}
body td,
body th {
padding: 0;
}
body * {
box-sizing: border-box;
}
body input {
font: 13px / 1.4 Helvetica, arial, nimbussansl, liberationsans, freesans, clean, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
}
body a {
color: #4078c0;
text-decoration: none;
}
body a:hover,
body a:active {
text-decoration: underline;
}
body hr {
height: 0;
margin: 15px 0;
overflow: hidden;
background: transparent;
border: 0;
border-bottom: 1px solid #ddd;
}
body hr:before {
display: table;
content: "";
}
body hr:after {
display: table;
clear: both;
content: "";
}
body h1,
body h2,
body h3,
body h4,
body h5,
body h6 {
margin-top: 15px;
margin-bottom: 15px;
line-height: 1.1;
}
body h1 {
font-size: 30px;
}
body h2 {
font-size: 21px;
}
body h3 {
font-size: 16px;
}
body h4 {
font-size: 14px;
}
body h5 {
font-size: 12px;
}
body h6 {
font-size: 11px;
}
body blockquote {
margin: 0;
}
body ul,
body ol {
padding: 0;
margin-top: 0;
margin-bottom: 0;
}
body ol ol,
body ul ol {
list-style-type: lower-roman;
}
body ul ul ol,
body ul ol ol,
body ol ul ol,
body ol ol ol {
list-style-type: lower-alpha;
}
body dd {
margin-left: 0;
}
body code {
font-family: Consolas, "Liberation Mono", Menlo, Courier, monospace;
font-size: 12px;
}
body pre {
margin-top: 0;
margin-bottom: 0;
font: 12px Consolas, "Liberation Mono", Menlo, Courier, monospace;
}
body .select::-ms-expand {
opacity: 0;
}
body .octicon {
font: normal normal normal 16px/1 octicons-link;
display: inline-block;
text-decoration: none;
text-rendering: auto;
-webkit-font-smoothing: antialiased;
-moz-osx-font-smoothing: grayscale;
-webkit-user-select: none;
-moz-user-select: none;
-ms-user-select: none;
user-select: none;
}
body .octicon-link:before {
content: '\f05c';
}
body:before {
display: table;
content: "";
}
body:after {
display: table;
clear: both;
content: "";
}
body>*:first-child {
margin-top: 0 !important;
}
body>*:last-child {
margin-bottom: 0 !important;
}
body a:not([href]) {
color: inherit;
text-decoration: none;
}
body .anchor {
display: inline-block;
padding-right: 2px;
margin-left: -18px;
}
body .anchor:focus {
outline: none;
}
body h1,
body h2,
body h3,
body h4,
body h5,
body h6 {
margin-top: 1em;
margin-bottom: 16px;
font-weight: bold;
line-height: 1.4;
}
body h1 .octicon-link,
body h2 .octicon-link,
body h3 .octicon-link,
body h4 .octicon-link,
body h5 .octicon-link,
body h6 .octicon-link {
color: #000;
vertical-align: middle;
visibility: hidden;
}
body h1:hover .anchor,
body h2:hover .anchor,
body h3:hover .anchor,
body h4:hover .anchor,
body h5:hover .anchor,
body h6:hover .anchor {
text-decoration: none;
}
body h1:hover .anchor .octicon-link,
body h2:hover .anchor .octicon-link,
body h3:hover .anchor .octicon-link,
body h4:hover .anchor .octicon-link,
body h5:hover .anchor .octicon-link,
body h6:hover .anchor .octicon-link {
visibility: visible;
}
body h1 {
padding-bottom: 0.3em;
font-size: 1.75em;
line-height: 1.2;
}
body h1 .anchor {
line-height: 1;
}
body h2 {
padding-bottom: 0.3em;
font-size: 1.5em;
line-height: 1.225;
}
body h2 .anchor {
line-height: 1;
}
body h3 {
font-size: 1.25em;
line-height: 1.43;
}
body h3 .anchor {
line-height: 1.2;
}
body h4 {
font-size: 1em;
}
body h4 .anchor {
line-height: 1.2;
}
body h5 {
font-size: 1em;
}
body h5 .anchor {
line-height: 1.1;
}
body h6 {
font-size: 1em;
color: #777;
}
body h6 .anchor {
line-height: 1.1;
}
body p,
body blockquote,
body ul,
body ol,
body dl,
body table,
body pre {
margin-top: 0;
margin-bottom: 16px;
}
body hr {
height: 4px;
padding: 0;
margin: 16px 0;
background-color: #e7e7e7;
border: 0 none;
}
body ul,
body ol {
padding-left: 2em;
}
body ul ul,
body ul ol,
body ol ol,
body ol ul {
margin-top: 0;
margin-bottom: 0;
}
body li>p {
margin-top: 16px;
}
body dl {
padding: 0;
}
body dl dt {
padding: 0;
margin-top: 16px;
font-size: 1em;
font-style: italic;
font-weight: bold;
}
body dl dd {
padding: 0 16px;
margin-bottom: 16px;
}
body blockquote {
padding: 0 15px;
color: #777;
border-left: 4px solid #ddd;
}
body blockquote>:first-child {
margin-top: 0;
}
body blockquote>:last-child {
margin-bottom: 0;
}
body table {
display: block;
width: 100%;
overflow: auto;
word-break: normal;
word-break: keep-all;
}
body table th {
font-weight: bold;
}
body table th,
body table td {
padding: 6px 13px;
border: 1px solid #ddd;
}
body table tr {
background-color: #fff;
border-top: 1px solid #ccc;
}
body table tr:nth-child(2n) {
background-color: #f8f8f8;
}
body img {
max-width: 100%;
box-sizing: content-box;
background-color: #fff;
}
body code {
padding: 0;
padding-top: 0.2em;
padding-bottom: 0.2em;
margin: 0;
font-size: 85%;
background-color: rgba(0,0,0,0.04);
border-radius: 3px;
}
body code:before,
body code:after {
letter-spacing: -0.2em;
content: "\00a0";
}
body pre>code {
padding: 0;
margin: 0;
font-size: 100%;
word-break: normal;
white-space: pre;
background: transparent;
border: 0;
}
body .highlight {
margin-bottom: 16px;
}
body .highlight pre,
body pre {
padding: 16px;
overflow: auto;
font-size: 85%;
line-height: 1.45;
background-color: #f7f7f7;
border-radius: 3px;
}
body .highlight pre {
margin-bottom: 0;
word-break: normal;
}
body pre {
word-wrap: normal;
}
body pre code {
display: inline;
max-width: initial;
padding: 0;
margin: 0;
overflow: initial;
line-height: inherit;
word-wrap: normal;
background-color: transparent;
border: 0;
}
body pre code:before,
body pre code:after {
content: normal;
}
body kbd {
display: inline-block;
padding: 3px 5px;
font-size: 11px;
line-height: 10px;
color: #555;
vertical-align: middle;
background-color: #fcfcfc;
border: solid 1px #ccc;
border-bottom-color: #bbb;
border-radius: 3px;
box-shadow: inset 0 -1px 0 #bbb;
}
body .pl-c {
color: #969896;
}
body .pl-c1,
body .pl-s .pl-v {
color: #0086b3;
}
body .pl-e,
body .pl-en {
color: #795da3;
}
body .pl-s .pl-s1,
body .pl-smi {
color: #333;
}
body .pl-ent {
color: #63a35c;
}
body .pl-k {
color: #a71d5d;
}
body .pl-pds,
body .pl-s,
body .pl-s .pl-pse .pl-s1,
body .pl-sr,
body .pl-sr .pl-cce,
body .pl-sr .pl-sra,
body .pl-sr .pl-sre {
color: #183691;
}
body .pl-v {
color: #ed6a43;
}
body .pl-id {
color: #b52a1d;
}
body .pl-ii {
background-color: #b52a1d;
color: #f8f8f8;
}
body .pl-sr .pl-cce {
color: #63a35c;
font-weight: bold;
}
body .pl-ml {
color: #693a17;
}
body .pl-mh,
body .pl-mh .pl-en,
body .pl-ms {
color: #1d3e81;
font-weight: bold;
}
body .pl-mq {
color: #008080;
}
body .pl-mi {
color: #333;
font-style: italic;
}
body .pl-mb {
color: #333;
font-weight: bold;
}
body .pl-md {
background-color: #ffecec;
color: #bd2c00;
}
body .pl-mi1 {
background-color: #eaffea;
color: #55a532;
}
body .pl-mdr {
color: #795da3;
font-weight: bold;
}
body .pl-mo {
color: #1d3e81;
}
body kbd {
display: inline-block;
padding: 3px 5px;
font: 11px Consolas, "Liberation Mono", Menlo, Courier, monospace;
line-height: 10px;
color: #555;
vertical-align: middle;
background-color: #fcfcfc;
border: solid 1px #ccc;
border-bottom-color: #bbb;
border-radius: 3px;
box-shadow: inset 0 -1px 0 #bbb;
}
body .task-list-item {
list-style-type: none;
}
body .task-list-item+.task-list-item {
margin-top: 3px;
}
body .task-list-item input {
margin: 0 0.35em 0.25em -1.6em;
vertical-align: middle;
}
body :checked+.radio-label {
z-index: 1;
position: relative;
border-color: #4078c0;
}
</style>
  <!--[if lt IE 9]>
    <script src="//cdnjs.cloudflare.com/ajax/libs/html5shiv/3.7.3/html5shiv-printshiv.min.js"></script>
  <![endif]-->
  
</head>
<body>
<nav id="TOC">
<ul>
<li><a href="#portalhail-api">Portal/Hail API</a><ul>
<li><a href="#api-specification">API specification</a><ul>
<li><a href="#retrieve-available-datasetsmasks">Retrieve available datasets/masks</a></li>
<li><a href="#retrieve-covariance-in-region-given-datasetmasks">Retrieve covariance in region given dataset/mask(s)</a></li>
<li><a href="#retrieving-pre-computed-aggregation-test-results">Retrieving pre-computed aggregation test results</a></li>
</ul></li>
</ul></li>
</ul>
</nav>
<h1 id="portalhail-api">Portal/Hail API</h1>
<h2 id="api-specification">API specification</h2>
<p>We propose a total of three endpoints. The first endpoint provides metadata that can be used to request a calculation.</p>
<p>The second and third endpoints are mutually exclusive: the former provides information needed to perform a calculation in the browser, while the latter provides pre-computed results. A helper function is provided to format live calculations to look like the final results- this allows both calculation methods to be used interchangeably and processed in the same way.</p>
<h3 id="retrieve-available-datasetsmasks">Retrieve available datasets/masks</h3>
<p>A genotype dataset is a set of variant genotypes and samples (e.g. a VCF.)</p>
<p>A phenotype dataset is a set of phenotypes and samples (PED or tab-delimited file). These phenotypes are linked to one or more genotype datasets.</p>
<p>A mask is defined by:</p>
<ol type="1">
<li>Variant filtering criteria</li>
<li>Variant grouping criteria</li>
</ol>
<p>Variant filtering:</p>
<ul>
<li>Allele frequency</li>
<li>Annotation about a variant (protein-truncating, loss-of-function, etc.)</li>
</ul>
<p>Grouping:</p>
<ul>
<li>Gene</li>
<li>Chromatin state regions (enhancer, silencer, promoter, insulator, etc.)</li>
<li>KEGG pathways (variants within genes that belong to the pathway)</li>
<li>Gene ontology terms (similar to KEGG)</li>
</ul>
<p>Masks would be available per genotype dataset.</p>
<p>For now, a description of the variant filtering and grouping criteria will probably just have to be plain text without some type of formal grammar.</p>
<h4 id="request">Request</h4>
<p><code>GET /api/aggregation/metadata</code></p>
<h4 id="response">Response</h4>
<div class="sourceCode" id="cb1"><pre class="sourceCode json"><code class="sourceCode json"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="fu">{</span></a>
<a class="sourceLine" id="cb1-2" data-line-number="2">  <span class="dt">&quot;data&quot;</span><span class="fu">:</span> <span class="ot">[</span></a>
<a class="sourceLine" id="cb1-3" data-line-number="3">    <span class="fu">{</span></a>
<a class="sourceLine" id="cb1-4" data-line-number="4">      <span class="dt">&quot;genotypeDataset&quot;</span><span class="fu">:</span> <span class="dv">1</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb1-5" data-line-number="5">      <span class="dt">&quot;description&quot;</span><span class="fu">:</span> <span class="st">&quot;52K Exomes&quot;</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb1-6" data-line-number="6">      <span class="dt">&quot;genomeBuild&quot;</span><span class="fu">:</span> <span class="st">&quot;GRCh37&quot;</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb1-7" data-line-number="7">      <span class="dt">&quot;masks&quot;</span><span class="fu">:</span> <span class="ot">[</span></a>
<a class="sourceLine" id="cb1-8" data-line-number="8">        <span class="fu">{</span></a>
<a class="sourceLine" id="cb1-9" data-line-number="9">          <span class="dt">&quot;id&quot;</span><span class="fu">:</span> <span class="dv">1</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb1-10" data-line-number="10">          <span class="dt">&quot;name&quot;</span><span class="fu">:</span> <span class="st">&quot;PTV&quot;</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb1-11" data-line-number="11">          <span class="dt">&quot;description&quot;</span><span class="fu">:</span> <span class="st">&quot;Protein truncating variants&quot;</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb1-12" data-line-number="12">          <span class="dt">&quot;groupType&quot;</span><span class="fu">:</span> <span class="st">&quot;GENE&quot;</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb1-13" data-line-number="13">          <span class="dt">&quot;identifierType&quot;</span><span class="fu">:</span> <span class="st">&quot;ENSEMBL&quot;</span></a>
<a class="sourceLine" id="cb1-14" data-line-number="14">        <span class="fu">}</span><span class="ot">,</span></a>
<a class="sourceLine" id="cb1-15" data-line-number="15">        <span class="fu">{</span></a>
<a class="sourceLine" id="cb1-16" data-line-number="16">          <span class="dt">&quot;id&quot;</span><span class="fu">:</span> <span class="dv">2</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb1-17" data-line-number="17">          <span class="dt">&quot;name&quot;</span><span class="fu">:</span> <span class="st">&quot;PTV &amp; LoF &amp; AF&lt;0.05&quot;</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb1-18" data-line-number="18">          <span class="dt">&quot;description&quot;</span><span class="fu">:</span> <span class="st">&quot;Protein truncating variants with AF &lt; 0.05&quot;</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb1-19" data-line-number="19">          <span class="dt">&quot;groupType&quot;</span><span class="fu">:</span> <span class="st">&quot;GENE&quot;</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb1-20" data-line-number="20">          <span class="dt">&quot;identifierType&quot;</span><span class="fu">:</span> <span class="st">&quot;ENSEMBL&quot;</span></a>
<a class="sourceLine" id="cb1-21" data-line-number="21">        <span class="fu">}</span><span class="ot">,</span></a>
<a class="sourceLine" id="cb1-22" data-line-number="22">        <span class="fu">{</span></a>
<a class="sourceLine" id="cb1-23" data-line-number="23">          <span class="dt">&quot;id&quot;</span><span class="fu">:</span> <span class="dv">3</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb1-24" data-line-number="24">          <span class="dt">&quot;name&quot;</span><span class="fu">:</span> <span class="st">&quot;PTV &amp; LoF &amp; AF&lt;0.05&quot;</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb1-25" data-line-number="25">          <span class="dt">&quot;description&quot;</span><span class="fu">:</span> <span class="st">&quot;Protein truncating and loss-of-function variants with AF &lt; 0.05&quot;</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb1-26" data-line-number="26">          <span class="dt">&quot;groupType&quot;</span><span class="fu">:</span> <span class="st">&quot;GENE&quot;</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb1-27" data-line-number="27">          <span class="dt">&quot;identifierType&quot;</span><span class="fu">:</span> <span class="st">&quot;ENSEMBL&quot;</span></a>
<a class="sourceLine" id="cb1-28" data-line-number="28">        <span class="fu">}</span></a>
<a class="sourceLine" id="cb1-29" data-line-number="29">      <span class="ot">]</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb1-30" data-line-number="30">      <span class="dt">&quot;phenotypeDatasets&quot;</span><span class="fu">:</span> <span class="ot">[</span></a>
<a class="sourceLine" id="cb1-31" data-line-number="31">        <span class="fu">{</span></a>
<a class="sourceLine" id="cb1-32" data-line-number="32">          <span class="dt">&quot;phenotypeDataset&quot;</span><span class="fu">:</span> <span class="dv">1</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb1-33" data-line-number="33">          <span class="dt">&quot;description&quot;</span><span class="fu">:</span> <span class="st">&quot;52K Exomes - Cardiovascular Traits&quot;</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb1-34" data-line-number="34">          <span class="dt">&quot;phenotypes&quot;</span><span class="fu">:</span> <span class="ot">[</span></a>
<a class="sourceLine" id="cb1-35" data-line-number="35">            <span class="fu">{</span></a>
<a class="sourceLine" id="cb1-36" data-line-number="36">              <span class="dt">&quot;name&quot;</span><span class="fu">:</span> <span class="st">&quot;ldl&quot;</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb1-37" data-line-number="37">              <span class="dt">&quot;description&quot;</span><span class="fu">:</span> <span class="st">&quot;Low density lipoproteins&quot;</span></a>
<a class="sourceLine" id="cb1-38" data-line-number="38">            <span class="fu">}</span><span class="ot">,</span></a>
<a class="sourceLine" id="cb1-39" data-line-number="39">            <span class="fu">{</span></a>
<a class="sourceLine" id="cb1-40" data-line-number="40">              <span class="dt">&quot;name&quot;</span><span class="fu">:</span> <span class="st">&quot;hdl&quot;</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb1-41" data-line-number="41">              <span class="dt">&quot;description&quot;</span><span class="fu">:</span> <span class="st">&quot;High density lipoproteins&quot;</span></a>
<a class="sourceLine" id="cb1-42" data-line-number="42">            <span class="fu">}</span></a>
<a class="sourceLine" id="cb1-43" data-line-number="43">          <span class="ot">]</span></a>
<a class="sourceLine" id="cb1-44" data-line-number="44">        <span class="fu">}</span><span class="ot">,</span></a>
<a class="sourceLine" id="cb1-45" data-line-number="45">        <span class="fu">{</span></a>
<a class="sourceLine" id="cb1-46" data-line-number="46">          <span class="dt">&quot;phenotypeDataset&quot;</span><span class="fu">:</span> <span class="dv">2</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb1-47" data-line-number="47">          <span class="dt">&quot;description&quot;</span><span class="fu">:</span> <span class="st">&quot;52K Exomes - Diabetes-related Traits&quot;</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb1-48" data-line-number="48">          <span class="dt">&quot;phenotypes&quot;</span><span class="fu">:</span> <span class="ot">[</span></a>
<a class="sourceLine" id="cb1-49" data-line-number="49">            <span class="fu">{</span></a>
<a class="sourceLine" id="cb1-50" data-line-number="50">              <span class="dt">&quot;name&quot;</span><span class="fu">:</span> <span class="st">&quot;T2D&quot;</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb1-51" data-line-number="51">              <span class="dt">&quot;description&quot;</span><span class="fu">:</span> <span class="st">&quot;Type 2 diabetes&quot;</span></a>
<a class="sourceLine" id="cb1-52" data-line-number="52">            <span class="fu">}</span><span class="ot">,</span></a>
<a class="sourceLine" id="cb1-53" data-line-number="53">            <span class="fu">{</span></a>
<a class="sourceLine" id="cb1-54" data-line-number="54">              <span class="dt">&quot;name&quot;</span><span class="fu">:</span> <span class="st">&quot;FG&quot;</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb1-55" data-line-number="55">              <span class="dt">&quot;description&quot;</span><span class="fu">:</span> <span class="st">&quot;Fasting glucose&quot;</span></a>
<a class="sourceLine" id="cb1-56" data-line-number="56">            <span class="fu">}</span></a>
<a class="sourceLine" id="cb1-57" data-line-number="57">          <span class="ot">]</span></a>
<a class="sourceLine" id="cb1-58" data-line-number="58">        <span class="fu">}</span></a>
<a class="sourceLine" id="cb1-59" data-line-number="59">      <span class="ot">]</span></a>
<a class="sourceLine" id="cb1-60" data-line-number="60">    <span class="fu">}</span></a>
<a class="sourceLine" id="cb1-61" data-line-number="61">  <span class="ot">]</span></a>
<a class="sourceLine" id="cb1-62" data-line-number="62"><span class="fu">}</span></a></code></pre></div>
<h3 id="retrieve-covariance-in-region-given-datasetmasks">Retrieve covariance in region given dataset/mask(s)</h3>
<h4 id="request-1">Request</h4>
<p>Only requesting scores and covariance for the “PTV” mask just to save some space.</p>
<p><code>POST /api/aggregation/covariance</code></p>
<div class="sourceCode" id="cb2"><pre class="sourceCode json"><code class="sourceCode json"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="fu">{</span></a>
<a class="sourceLine" id="cb2-2" data-line-number="2">  <span class="dt">&quot;chrom&quot;</span><span class="fu">:</span> <span class="st">&quot;6&quot;</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb2-3" data-line-number="3">  <span class="dt">&quot;start&quot;</span><span class="fu">:</span> <span class="dv">1</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb2-4" data-line-number="4">  <span class="dt">&quot;stop&quot;</span><span class="fu">:</span> <span class="dv">100000</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb2-5" data-line-number="5">  <span class="dt">&quot;genotypeDataset&quot;</span><span class="fu">:</span> <span class="dv">1</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb2-6" data-line-number="6">  <span class="dt">&quot;phenotypeDataset&quot;</span><span class="fu">:</span> <span class="dv">1</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb2-7" data-line-number="7">  <span class="dt">&quot;phenotype&quot;</span><span class="fu">:</span> <span class="st">&quot;rand_qt&quot;</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb2-8" data-line-number="8">  <span class="dt">&quot;samples&quot;</span><span class="fu">:</span> <span class="st">&quot;ALL&quot;</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb2-9" data-line-number="9">  <span class="dt">&quot;genomeBuild&quot;</span><span class="fu">:</span> <span class="st">&quot;GRCh37&quot;</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb2-10" data-line-number="10">  <span class="dt">&quot;masks&quot;</span><span class="fu">:</span> <span class="ot">[</span><span class="er">...</span><span class="ot">]</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb2-11" data-line-number="11">  <span class="dt">&quot;maskDefinitions&quot;</span><span class="fu">:</span> <span class="ot">[</span><span class="er">...</span><span class="ot">]</span></a>
<a class="sourceLine" id="cb2-12" data-line-number="12"><span class="fu">}</span></a></code></pre></div>
<p>The <code>masks</code> should be a list of integer IDs, representing server-side mask files (for example, if a mask was already computed for a particular genotype dataset). For example:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode json"><code class="sourceCode json"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="fu">{</span></a>
<a class="sourceLine" id="cb3-2" data-line-number="2">  <span class="dt">&quot;masks&quot;</span><span class="fu">:</span> <span class="ot">[</span><span class="dv">1</span><span class="ot">,</span> <span class="dv">2</span><span class="ot">,</span> <span class="dv">3</span><span class="ot">]</span></a>
<a class="sourceLine" id="cb3-3" data-line-number="3"><span class="fu">}</span></a></code></pre></div>
<p>You can instead supply a list of your own generated masks in the browser or client-side using the <code>maskDefinitions</code> key:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode json"><code class="sourceCode json"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="fu">{</span></a>
<a class="sourceLine" id="cb4-2" data-line-number="2">  <span class="dt">&quot;maskDefinitions&quot;</span><span class="fu">:</span> <span class="ot">[</span></a>
<a class="sourceLine" id="cb4-3" data-line-number="3">    <span class="fu">{</span></a>
<a class="sourceLine" id="cb4-4" data-line-number="4">      <span class="dt">&quot;id&quot;</span><span class="fu">:</span> <span class="dv">10</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb4-5" data-line-number="5">      <span class="dt">&quot;name&quot;</span><span class="fu">:</span> <span class="st">&quot;PTV+LOF&lt;0.01&quot;</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb4-6" data-line-number="6">      <span class="dt">&quot;description&quot;</span><span class="fu">:</span> <span class="st">&quot;A mask generated by the user in the browser, including only variants that are PTV or LOF w/ AF &lt; 1%&quot;</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb4-7" data-line-number="7">      <span class="dt">&quot;genome_build&quot;</span><span class="fu">:</span> <span class="st">&quot;GRCh37&quot;</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb4-8" data-line-number="8">      <span class="dt">&quot;group_type&quot;</span><span class="fu">:</span> <span class="st">&quot;GENE&quot;</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb4-9" data-line-number="9">      <span class="dt">&quot;identifier_type&quot;</span><span class="fu">:</span> <span class="st">&quot;ENSEMBL&quot;</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb4-10" data-line-number="10">      <span class="dt">&quot;groups&quot;</span><span class="fu">:</span> <span class="fu">{</span></a>
<a class="sourceLine" id="cb4-11" data-line-number="11">        <span class="dt">&quot;ENSG000001&quot;</span><span class="fu">:</span> <span class="ot">[</span><span class="st">&quot;1:1_A/G&quot;</span><span class="ot">,</span> <span class="st">&quot;1:2_C/T&quot;</span><span class="ot">,</span> <span class="er">...</span><span class="ot">]</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb4-12" data-line-number="12">        <span class="dt">&quot;ENSG000002&quot;</span><span class="fu">:</span> <span class="ot">[</span><span class="st">&quot;1:43_G/C&quot;</span><span class="ot">,</span> <span class="er">...</span><span class="ot">]</span></a>
<a class="sourceLine" id="cb4-13" data-line-number="13">      <span class="fu">}</span></a>
<a class="sourceLine" id="cb4-14" data-line-number="14">    <span class="fu">}</span></a>
<a class="sourceLine" id="cb4-15" data-line-number="15">  <span class="ot">]</span></a>
<a class="sourceLine" id="cb4-16" data-line-number="16"><span class="fu">}</span></a></code></pre></div>
<p>In this case, the ID can be any integer, so long as it is unique across all of the masks sent in the request.</p>
<p>Note that both <code>masks</code> and <code>maskDefinitions</code> should not be given together.</p>
<h4 id="response-1">Response</h4>
<p>We want to be able to retrieve as much of the masks/covariance data all at once to minimize round trips to the server. This structure would allow retrieving covariance for multiple masks/groups all at once.</p>
<p>Note that all scores and covariances are assumed to be counting towards the <strong>alternate allele</strong>. It is also critical that alternate allele frequencies are included so as to be able to orient scores/covariances towards the minor allele when performing the aggregation tests.</p>
<p>The data section of this payload contains two distinct kinds of entities:</p>
<ul>
<li><p><strong>groups</strong> represent the information required to perform analysis on one or more sets of variants. Often, these variants are based on a particular region (such as <code>groupType: 'interval'</code> or <code>groupType: 'gene'</code>), filtered according to a specific mask. In the example below, only one set is specified: all protein truncating variantsa (PTV) that lie within the region that defines a gene with ENSEMBL ID <code>ENSG000001</code>.</p></li>
<li><p><strong>variants</strong> contains basic information about individual variants. Some information (such as <code>altFreq</code>) is needed for calculations. Optionally, this section may present additional information that can be used to analyze and interpret the calculation results. The same variant may appear in many groups, but is only listed here once.</p></li>
</ul>
<div class="sourceCode" id="cb5"><pre class="sourceCode json"><code class="sourceCode json"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="fu">{</span></a>
<a class="sourceLine" id="cb5-2" data-line-number="2">  <span class="dt">&quot;data&quot;</span><span class="fu">:</span> <span class="fu">{</span></a>
<a class="sourceLine" id="cb5-3" data-line-number="3">    <span class="dt">&quot;genotypeDataset&quot;</span><span class="fu">:</span> <span class="dv">42</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb5-4" data-line-number="4">    <span class="dt">&quot;phenotypeDataset&quot;</span><span class="fu">:</span> <span class="dv">8</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb5-5" data-line-number="5">    <span class="dt">&quot;phenotype&quot;</span><span class="fu">:</span> <span class="st">&quot;T2D&quot;</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb5-6" data-line-number="6">    <span class="dt">&quot;sigmaSquared&quot;</span><span class="fu">:</span> <span class="fl">0.08</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb5-7" data-line-number="7">    <span class="dt">&quot;nSamples&quot;</span><span class="fu">:</span> <span class="dv">3550</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb5-8" data-line-number="8">    <span class="dt">&quot;variants&quot;</span><span class="fu">:</span> <span class="ot">[</span></a>
<a class="sourceLine" id="cb5-9" data-line-number="9">      <span class="fu">{</span></a>
<a class="sourceLine" id="cb5-10" data-line-number="10">        <span class="dt">&quot;variant&quot;</span><span class="fu">:</span> <span class="st">&quot;2:21228642_G/A&quot;</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb5-11" data-line-number="11">        <span class="dt">&quot;altFreq&quot;</span><span class="fu">:</span> <span class="fl">0.033</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb5-12" data-line-number="12">        <span class="dt">&quot;pvalue&quot;</span><span class="fu">:</span> <span class="fl">0.000431</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb5-13" data-line-number="13">        <span class="dt">&quot;score&quot;</span><span class="fu">:</span> <span class="fl">0.1</span></a>
<a class="sourceLine" id="cb5-14" data-line-number="14">      <span class="fu">}</span></a>
<a class="sourceLine" id="cb5-15" data-line-number="15">    <span class="ot">]</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb5-16" data-line-number="16">    <span class="dt">&quot;groups&quot;</span><span class="fu">:</span> <span class="ot">[</span></a>
<a class="sourceLine" id="cb5-17" data-line-number="17">      <span class="fu">{</span></a>
<a class="sourceLine" id="cb5-18" data-line-number="18">        <span class="dt">&quot;groupType&quot;</span><span class="fu">:</span> <span class="st">&quot;GENE&quot;</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb5-19" data-line-number="19">        <span class="dt">&quot;group&quot;</span><span class="fu">:</span> <span class="st">&quot;ENSG000001&quot;</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb5-20" data-line-number="20">        <span class="dt">&quot;mask&quot;</span><span class="fu">:</span> <span class="dv">1</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb5-21" data-line-number="21">        <span class="dt">&quot;variants&quot;</span><span class="fu">:</span> <span class="ot">[</span><span class="st">&quot;2:21228642_G/A&quot;</span><span class="ot">]</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb5-22" data-line-number="22">        <span class="dt">&quot;covariance&quot;</span><span class="fu">:</span> <span class="ot">[</span><span class="fl">0.3</span><span class="ot">]</span></a>
<a class="sourceLine" id="cb5-23" data-line-number="23">      <span class="fu">}</span></a>
<a class="sourceLine" id="cb5-24" data-line-number="24">    <span class="ot">]</span></a>
<a class="sourceLine" id="cb5-25" data-line-number="25">  <span class="fu">}</span></a>
<a class="sourceLine" id="cb5-26" data-line-number="26"><span class="fu">}</span></a></code></pre></div>
<h3 id="retrieving-pre-computed-aggregation-test-results">Retrieving pre-computed aggregation test results</h3>
<p>In the case where aggregation tests have already been pre-computed for some of the datasets server-side, this endpoint provides access to calculation results (and associated information). This endpoint is mutually exclusive with the one above: if results have already been computed, then there is no need to retrieve the raw data (such as covariance) used as input to the calculation.</p>
<p>Otherwise, the payloads are superficially similar, so that the same visualization experience can work with both on-the-fly and precomputed results. <code>raremetal.js</code> has a helper function that will format in-browser calculations as the response below.</p>
<h4 id="request-2">Request</h4>
<p>Looks identical to the same request used to retrieve covariance.</p>
<p><code>POST /api/aggregation/results</code></p>
<div class="sourceCode" id="cb6"><pre class="sourceCode json"><code class="sourceCode json"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="fu">{</span></a>
<a class="sourceLine" id="cb6-2" data-line-number="2">  <span class="dt">&quot;chrom&quot;</span><span class="fu">:</span> <span class="st">&quot;6&quot;</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb6-3" data-line-number="3">  <span class="dt">&quot;start&quot;</span><span class="fu">:</span> <span class="dv">1</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb6-4" data-line-number="4">  <span class="dt">&quot;stop&quot;</span><span class="fu">:</span> <span class="dv">100000</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb6-5" data-line-number="5">  <span class="dt">&quot;genotypeDataset&quot;</span><span class="fu">:</span> <span class="dv">1</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb6-6" data-line-number="6">  <span class="dt">&quot;phenotypeDataset&quot;</span><span class="fu">:</span> <span class="dv">1</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb6-7" data-line-number="7">  <span class="dt">&quot;phenotype&quot;</span><span class="fu">:</span> <span class="st">&quot;rand_qt&quot;</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb6-8" data-line-number="8">  <span class="dt">&quot;samples&quot;</span><span class="fu">:</span> <span class="st">&quot;ALL&quot;</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb6-9" data-line-number="9">  <span class="dt">&quot;genomeBuild&quot;</span><span class="fu">:</span> <span class="st">&quot;GRCh37&quot;</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb6-10" data-line-number="10">  <span class="dt">&quot;masks&quot;</span><span class="fu">:</span> <span class="ot">[</span></a>
<a class="sourceLine" id="cb6-11" data-line-number="11">    <span class="dv">1</span></a>
<a class="sourceLine" id="cb6-12" data-line-number="12">  <span class="ot">]</span></a>
<a class="sourceLine" id="cb6-13" data-line-number="13"><span class="fu">}</span></a></code></pre></div>
<h4 id="response-2">Response</h4>
<p>If results were already available server-side, say if they were pre-computed, we could accept a response of this format and be able to show the results within LZ without performing any aggregation test computations.</p>
<p>The two types of data in this endpoint are similar to those described in the <code>covariance</code> endpoint above. Instead of covariance data, <code>groups</code> in this endpoint provide calculation results.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode json"><code class="sourceCode json"><a class="sourceLine" id="cb7-1" data-line-number="1"><span class="fu">{</span></a>
<a class="sourceLine" id="cb7-2" data-line-number="2">  <span class="dt">&quot;data&quot;</span><span class="fu">:</span> <span class="fu">{</span></a>
<a class="sourceLine" id="cb7-3" data-line-number="3">    <span class="dt">&quot;genotypeDataset&quot;</span><span class="fu">:</span> <span class="dv">42</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb7-4" data-line-number="4">    <span class="dt">&quot;description&quot;</span><span class="fu">:</span> <span class="st">&quot;52K Exomes&quot;</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb7-5" data-line-number="5">    <span class="dt">&quot;phenotypeDataset&quot;</span><span class="fu">:</span> <span class="dv">1</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb7-6" data-line-number="6">    <span class="dt">&quot;phenotype&quot;</span><span class="fu">:</span> <span class="st">&quot;T2D&quot;</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb7-7" data-line-number="7">    <span class="dt">&quot;variants&quot;</span><span class="fu">:</span> <span class="ot">[</span></a>
<a class="sourceLine" id="cb7-8" data-line-number="8">      <span class="fu">{</span></a>
<a class="sourceLine" id="cb7-9" data-line-number="9">        <span class="dt">&quot;variant&quot;</span><span class="fu">:</span> <span class="st">&quot;2:21228642_G/A&quot;</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb7-10" data-line-number="10">        <span class="dt">&quot;altFreq&quot;</span><span class="fu">:</span> <span class="fl">0.033</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb7-11" data-line-number="11">        <span class="dt">&quot;pvalue&quot;</span><span class="fu">:</span> <span class="fl">0.000431</span></a>
<a class="sourceLine" id="cb7-12" data-line-number="12">      <span class="fu">}</span></a>
<a class="sourceLine" id="cb7-13" data-line-number="13">    <span class="ot">]</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb7-14" data-line-number="14">    <span class="dt">&quot;groups&quot;</span><span class="fu">:</span> <span class="ot">[</span></a>
<a class="sourceLine" id="cb7-15" data-line-number="15">      <span class="fu">{</span></a>
<a class="sourceLine" id="cb7-16" data-line-number="16">        <span class="dt">&quot;groupType&quot;</span><span class="fu">:</span> <span class="st">&quot;GENE&quot;</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb7-17" data-line-number="17">        <span class="dt">&quot;group&quot;</span><span class="fu">:</span> <span class="st">&quot;ENSG000001&quot;</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb7-18" data-line-number="18">        <span class="dt">&quot;mask&quot;</span><span class="fu">:</span> <span class="st">&quot;PTV&quot;</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb7-19" data-line-number="19">        <span class="dt">&quot;variants&quot;</span><span class="fu">:</span> <span class="ot">[</span><span class="st">&quot;2:21228642_G/A&quot;</span><span class="ot">]</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb7-20" data-line-number="20">        <span class="dt">&quot;test&quot;</span><span class="fu">:</span> <span class="st">&quot;burden&quot;</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb7-21" data-line-number="21">        <span class="dt">&quot;pvalue&quot;</span><span class="fu">:</span> <span class="fl">1.8e-09</span><span class="fu">,</span></a>
<a class="sourceLine" id="cb7-22" data-line-number="22">        <span class="dt">&quot;stat&quot;</span><span class="fu">:</span> <span class="fl">0.1</span></a>
<a class="sourceLine" id="cb7-23" data-line-number="23">      <span class="fu">}</span></a>
<a class="sourceLine" id="cb7-24" data-line-number="24">    <span class="ot">]</span></a>
<a class="sourceLine" id="cb7-25" data-line-number="25">  <span class="fu">}</span></a>
<a class="sourceLine" id="cb7-26" data-line-number="26"><span class="fu">}</span></a></code></pre></div>
</body>
</html>
